lib/bamboo/test.ex

```diff
@@ -90,23 +94,91 @@ defmodule Bamboo.Test do
       assert_delivered_email(unsent_email) # Will fail
   """
   def assert_delivered_email(%Bamboo.Email{} = email) do
-    import ExUnit.Assertions
     email = Bamboo.Mailer.normalize_addresses(email)
-    assert_received {:delivered_email, ^email}
+
+    do_assert_delivered_email(email)
+  end
+
+  defp do_assert_delivered_email(email) do
+    receive do
+      {:delivered_email, ^email} -> true
+    after
+      @timeout -> flunk_with_email_list(email)
+    end
+  end
+
+  defp flunk_with_email_list(email) do
+    if Enum.empty?(delivered_emails) do
+      flunk """
+      There were 0 emails delivered to this process.
+
+      If you expected an email to be sent, try these ideas:
+
+        1) Make sure you call deliver_now/1 or deliver_later/1 to deliver the email
+        2) Make sure you are using the Bamboo.TestAdapter
+        3) Use the process name feature of Bamboo.Test. This will allow Bamboo.Test
+           to work across processes: use Bamboo.Test, process_name: :my_test_name
+        4) If you are writing an acceptance test through a headless browser, see
+           if writing a controller test would be possible instead. Otherwise try
+           option 3.
+      """
+    else
+      flunk """
+      There were no matching emails.
+
+      No emails matched:
+
+        #{inspect email}
+
+      Delivered emails:
+
+      #{delivered_emails_as_list}
+      """
+    end
   end
-  def assert_delivered_email(email_options) when is_list(email_options) do
-    import ExUnit.Assertions
-    email = Bamboo.Email.new_email(email_options)
-      |> Bamboo.Mailer.normalize_addresses
-    assert_received {:delivered_email, ^email}
+
+  defp delivered_emails do
+    {:messages, messages} = Process.info(self, :messages)
+
+    for {:delivered_email, _} = email_message <- messages do
+      email_message
+    end
+  end
+
+  defp delivered_emails_as_list do
+    delivered_emails |> add_asterisk |> Enum.join("\n")
+  end
+
+  defp add_asterisk(emails) do
+    for email <- emails do
+      "  * #{inspect email}"
+    end
   end
 
   @doc """
   Checks that no emails were sent.
   """
+  def assert_no_emails_delivered do
+    receive do
+      {:delivered_email, email} -> flunk_with_unexpected_email(email)
+    after
+      0 -> true
```

Comment:
Before I merge this in I want to wait for #105 to close. Then I can raise with a message that tells people to set a `refute_delivery_timeout` that is required when using `process_name` with Bamboo.Test. That way for most tests it has 0 timeout, but for process_name people can choose how long they wait